hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/s3/tenant/OMTenantCreateRequest.java

```diff
@@ -0,0 +1,300 @@
+      throw new OMException("Invalid tenant name " + tenantName +
+          ". Tenant name should not contain delimiter.",
+          OMException.ResultCodes.INVALID_VOLUME_NAME);
+    }
+
+    // getUserName returns:
+    // - Kerberos principal when Kerberos security is enabled
+    // - User's login name when security is not enabled
+    // - AWS_ACCESS_KEY_ID if the original request comes from S3 Gateway.
+    //    Not Applicable to TenantCreateRequest.
+    final String owner = getOmRequest().getUserInfo().getUserName();
+    final String volumeName = tenantName;  // TODO: Configurable
+    final VolumeInfo volumeInfo = VolumeInfo.newBuilder()
+        .setVolume(volumeName)
+        .setAdminName(owner)
+        .setOwnerName(owner)
+        .build();
+    // Verify volume name
+    OmUtils.validateVolumeName(volumeInfo.getVolume());
+
+    // Generate volume modification time
+    long initialTime = Time.now();
+    final VolumeInfo updatedVolumeInfo = volumeInfo.toBuilder()
+            .setCreationTime(initialTime)
+            .setModificationTime(initialTime)
+            .build();
+
+    final OMRequest.Builder omRequestBuilder = getOmRequest().toBuilder()
+        .setCreateTenantRequest(
+            CreateTenantRequest.newBuilder().setTenantName(tenantName))
+        .setCreateVolumeRequest(
+            CreateVolumeRequest.newBuilder().setVolumeInfo(updatedVolumeInfo))
+        .setUserInfo(getUserInfo())
+        .setCmdType(getOmRequest().getCmdType())
+        .setClientId(getOmRequest().getClientId());
+
+    if (getOmRequest().hasTraceID()) {
+      omRequestBuilder.setTraceID(getOmRequest().getTraceID());
+    }
+
+    return omRequestBuilder.build();
+  }
+
+  @Override
+  public OMClientResponse validateAndUpdateCache(
+      OzoneManager ozoneManager, long transactionLogIndex,
+      OzoneManagerDoubleBufferHelper ozoneManagerDoubleBufferHelper) {
+
+    OMClientResponse omClientResponse = null;
+    OMResponse.Builder omResponse = OmResponseUtil.getOMResponseBuilder(
+        getOmRequest());
+    OmVolumeArgs omVolumeArgs;
+    boolean acquiredVolumeLock = false, acquiredUserLock = false;
+    Tenant tenant = null;
+    final String owner = getOmRequest().getUserInfo().getUserName();
+    LOG.info("owner: {}", owner);  // TODO: owner should be access_key_id
+    Map<String, String> auditMap = new HashMap<>();
+    OMMetadataManager omMetadataManager = ozoneManager.getMetadataManager();
+    CreateTenantRequest request = getOmRequest().getCreateTenantRequest();
+    final String tenantName = request.getTenantName();
+    final VolumeInfo volumeInfo =
+        getOmRequest().getCreateVolumeRequest().getVolumeInfo();
+    final String volumeName = volumeInfo.getVolume();
+    final String dbVolumeKey = omMetadataManager.getVolumeKey(volumeName);
+    IOException exception = null;
+    try {
+      // Check ACL: requires volume create permission. TODO: tenant create perm?
```

Comment:
Can any user who has volume create permissions on Ranger (or admin) be allowed to create a tenant?

Reply:
With the current logic, yes.